x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java

```diff
@@ -44,23 +53,68 @@ public class EnableSpatialDistancePushdown extends PhysicalOptimizerRules.Parame
     @Override
     protected PhysicalPlan rule(FilterExec filterExec, LocalPhysicalOptimizerContext ctx) {
         PhysicalPlan plan = filterExec;
-        if (filterExec.child() instanceof EsQueryExec) {
+        if (filterExec.child() instanceof EsQueryExec esQueryExec) {
+            plan = rewrite(filterExec, esQueryExec);
+        } else if (filterExec.child() instanceof EvalExec evalExec && evalExec.child() instanceof EsQueryExec esQueryExec) {
+            plan = rewrite(filterExec, evalExec, esQueryExec);
+        }
+
+        return plan;
+    }
+
+    private FilterExec rewrite(FilterExec filterExec, EsQueryExec esQueryExec) {
+        // Find and rewrite any binary comparisons that involve a distance function and a literal
+        var rewritten = filterExec.condition().transformDown(EsqlBinaryComparison.class, comparison -> {
+            ComparisonType comparisonType = ComparisonType.from(comparison.getFunctionType());
+            if (comparison.left() instanceof StDistance dist && comparison.right().foldable()) {
+                return rewriteComparison(comparison, dist, comparison.right(), comparisonType);
+            } else if (comparison.right() instanceof StDistance dist && comparison.left().foldable()) {
+                return rewriteComparison(comparison, dist, comparison.left(), ComparisonType.invert(comparisonType));
+            }
+            return comparison;
+        });
+        if (rewritten.equals(filterExec.condition()) == false) {
+            return new FilterExec(filterExec.source(), esQueryExec, rewritten);
+        }
+        return filterExec;
+    }
+
+    private Map<NameId, StDistance> getPushableDistances(List<Alias> aliases) {
+        LinkedHashMap<NameId, StDistance> refs = new LinkedHashMap<>();
```

Comment:
Normally we use AttributeMap for this (it's also just a map based on name id equality), but that's also fine.